### PR TITLE
Register species from MICM

### LINKED ISSUE #74: Register chemical species from MICM wrapper
Allow MICM to tell the host model what chemical species should be included in the state and provide their chemical/physical properties.

**Acceptance Criteria**
- The set of chemical species and their molecular weights from a MICM configuration are communicated to CAM-SIMA at run-time to set the model's constituents array.
- The correct species concentrations are updated during the "run" phase in the MICM wrapper.

**Ideas**
- Create a `micm_register` function that initially will be called directly by CAM-SIMA to set the constituents array. (Eventually, this register function may become a standard feature of the CCPP framework)
- The `micm_register` function should return an array of type `ccpp_constituent_properties_t` that will be used as CAM-SIMA sets its comprehensive set of constituents [here](https://github.com/ESCOMP/CAM-SIMA/blob/7d299b6671dcac75d236dd52be2abc73f679c5eb/src/control/cam_comp.F90#L56)

### PR DESCRIPTION
**Pull request by mattldawson**
Registers constituents and sets needed constituent properties from MICM configuration.

Additionally:
- returns error codes and messages from MUSICA functions using new error handling in MICM
- creates a `musica_ccpp` module to act as a the CCPP interface and moves existing functionality to a `musica_ccpp_micm` module in preparation for adding TUV-x (we will use `musica_ccpp_` as a prefix for all the MUSICA modules in this repo to prevent naming conflicts with MUSICA Fortran library modules)
- updates build scripts for changes to MUSICA and CCPP framework
- adds memory checking to MUSICA tests

closes #74

### CODE REVIEW

**Review by K20shores — APPROVED** [review]
I like the change to a `musica_ccpp_*` interface

File: musica/micm/micm.F90
Comment (mattldawson) [reply musica/micm/micm.F90]: good point. I just changed everything to `kind_phys` to be consistent

File: test/docker/Dockerfile.musica
Comment (mattldawson) [reply test/docker/Dockerfile.musica]: done!

File: musica/musica.meta
Comment (cacraigucar) [inline musica/musica.meta:7]: I think you need a musica_register metadata section

File: musica/util.F90
Comment (cacraigucar) [inline musica/util.F90:38]: github is indicating that there is no "newline" at the end of this file.   stackoverlow indicates that best practice is a newline should be present on the last line of a file.

**Review by cacraigucar — CHANGES_REQUESTED** [review]
I've reviewed from a Fortran standpoint.  @nusbaume and @peverwhee will need to review the interactions with the constituent object and the testing sections as I've not attempted to review them.

File: musica/util.F90
Comment (mattldawson) [reply musica/util.F90]: done!

File: musica/micm/micm.F90
Comment (mattldawson) [reply musica/micm/micm.F90]: done!

File: musica/musica.meta
Comment (mattldawson) [reply musica/musica.meta:7]: I believe for the dynamic species function, the typical meta data is replaced by this line:
https://github.com/mattldawson/atmospheric_physics/blob/6fb36e8b6356af8e50ab0df1a6b8afd0f4f29318/musica/musica.meta#L4

but I will defer to @peverwhee on this

**Review by nusbaume — CHANGES_REQUESTED** [review]
Thanks for implementing the new register phase and memory test!  However I do have some suggestions and questions, including a few questions which may (?) impact the overall design.  Happy to discuss anything in more detail if you'd like!

Also, just an FYI that (at least for now) the PR will require and update to the ChangeLog before being fully merged:

https://github.com/ESCOMP/atmospheric_physics/wiki/Development-workflow#changelog
 
Thanks again!

File: musica/micm/micm.F90
Comment (nusbaume) [reply musica/micm/micm.F90:9]: So I realized this morning that you might have done it this way because you can't pass in extra variables into the register call (is that true @peverwhee)?  If so then I think keeping this `use` statement here is totally fine for this PR, but we on the CCPP-side will need to figure out a way to pass in configuration files into register phase subroutines like this one soon(-ish).

File: test/musica/musica_namelist.F90
Comment (nusbaume) [reply test/musica/musica_namelist.F90:9]: Actually if the current constituent register call prevents the addition of an extra subroutine input argument then feel free to ignore this request, at least until we figure out a way to allow for that setup on the CCPP-framework side (which will certainly be after this PR).  Given this I'll go ahead and resolve this particular request.

File: musica/micm/micm.F90
Comment (peverwhee) [reply musica/micm/micm.F90:9]: @nusbaume - correct. we don't currently have metadata for the dynamic constituent routines.

File: musica/musica.meta
Comment (peverwhee) [reply musica/musica.meta:51]: It is unfortunately case-sensitive right now until we get https://github.com/NCAR/ccpp-framework/issues/567 fixed on the framework side.

File: musica/musica.meta
Comment (peverwhee) [reply musica/musica.meta:4]: I believe @mattldawson has added the dependencies line in his sandbox, but CAM-SIMA doesn't currently do anything with the dependencies (i.e. grab them during the build)

I've added an agenda item for tuesday's meeting.

Comment (nusbaume) [reply musica/musica.meta:4]: Thanks for letting me know @peverwhee!  Happy to discuss this Tuesday, although I am guessing if the framework is able to provide the list of dependency files at code generation time it will probably be trivial (e.g. a handful of new lines) to get the correct info into CAM-SIMA for successful building/linking.

File: musica/micm/micm.F90
Comment (nusbaume) [reply musica/micm/micm.F90:9]: @peverwhee Thanks for the confirmation!  In that case I'll go ahead and mark this concern resolved (as I don't think we'll want to change it for this PR).  However, I made a note for our Wednesday meeting to discuss it, as it is definitely a capability I think we'll need long-term.

File: suite_musica.xml
Comment (nusbaume) [reply suite_musica.xml:6]: After discussing this at the CAM-SIMA meeting we decided to just leave this interface as-is, at least for now, so I will go ahead and mark this comment resolved.

File: musica/micm/micm.F90
Comment (mattldawson) [reply musica/micm/micm.F90:88]: I typically try to limit the scope of dependencies (i.e. only use something at the module level if it's needed for a module variable or type definition). I've found that this makes the code more readable, but this is only my personal preference. I would suggest though that we either move everything to the module level or keep use statements at the minimum scope possible, just to make it easier to know where to look for things.

Comment (gold2718) [reply musica/micm/micm.F90:88]: I'm a fan of the limited scope style, mostly for the reason @mattldawson mentions.
@nusbaume, is moving to module scope your personal preference or something the AMP SEs would like to see more (e.g., potentially as a [coding rule](https://github.com/ESCOMP/CAM/wiki/CAM-Coding-Standards#fortran-coding-standards))?

File: musica/micm/micm.F90
Comment (mattldawson) [reply musica/micm/micm.F90]: Yes, this should have used `num_layers` as the second dimension (updated). The last dimension will be the number of externally supplied rate constants (e.g., photolysis rate constants), but since we don't have any of these yet, I left the last dimension set to `0`.

File: musica/musica.F90
Comment (mattldawson) [reply musica/musica.F90:13]: (see above comment)

File: musica/musica.F90
Comment (mattldawson) [reply musica/musica.F90]: removed!

File: test/docker/Dockerfile.musica
Comment (mattldawson) [reply test/docker/Dockerfile.musica]: This doesn't seem to have been used, so I removed it. If we need it in the future we can reintroduce it.

File: test/musica/test_musica_api.F90
Comment (mattldawson) [reply test/musica/test_musica_api.F90]: Absolutely! Added checks for the molecular weight and advected flag

File: musica/musica.meta
Comment (mattldawson) [reply musica/musica.meta:4]: After our discussion in the CAM-SIMA development meeting, I believe the plan is to wait until the constituents PR for the CCPP-framework is merged in (https://github.com/NCAR/ccpp-framework/pull/549) and then create an issue to update the framework to provide absolute paths for dependencies in the `ccpp_datatable.xml` file. This will allow CAM-SIMA to include the dependencies in the build with minor modification. @nusbaume @peverwhee - is this correct?

File: musica/util.F90
Comment (cacraigucar) [reply musica/util.F90]: This solved the problem for this particular file.  The newline issue occurs on most (but not all) of the files in this PR.  Let me know if you don't see the red circle with the dash in it, and I can indicated each file with the problem

Comment (mattldawson) [reply musica/util.F90]: I think they should all now have newlines at the end. Let me know if I missed any!

**Review by cacraigucar — APPROVED** [review]
@mattldawson - you found all the files with the missing newlines!

File: musica/micm/micm.F90
Comment (nusbaume) [reply musica/micm/micm.F90:88]: Hah, well you all have put more thought into this then I have!  After discussing it with the other AMP SEs we agreed that bringing `use` statements in at the smallest scope possible was a good idea, so I think this organization is good as-is (and will resolve this suggestion).  I also added a new rule to the CAM Fortran coding standards, which you can find at the bottom of the "SHOULD" list here:

 https://github.com/ESCOMP/CAM/wiki/CAM-Coding-Standards#should-1

File: musica/musica.meta
Comment (nusbaume) [reply musica/musica.meta:4]: Yep, that sounds like the correct plan to me!

**Review by nusbaume — COMMENTED** [review]
@mattldawson this PR looks good to me!  However I'll hold off on approving until we can get it working in CAM-SIMA (i.e. fix the `dependencies` issue in the framework).  Of course if you'd like this PR in sooner than that just let me know.  Thanks!

**Comment by mattldawson** [comment]
> @mattldawson this PR looks good to me! However I'll hold off on approving until we can get it working in CAM-SIMA (i.e. fix the `dependencies` issue in the framework). Of course if you'd like this PR in sooner than that just let me know. Thanks!

That's fine by me. There are some TUV-x-related developments in the pipeline, but hopefully the CAM-SIMA updates will be in before these are ready.

**Comment by mattldawson** [comment]
closing and reopening on `development` branch

**Pull request closed without merging**
